### PR TITLE
Sovchinnik erlcloud ec2 pagination feature

### PR DESCRIPTION
**Pull request by SergeyOvchinnik**
Problem: 
Function describe_instances in erlcloud_ec2 did not support pagination feature on library level

Solution:
Implemented pagination feature for erlcloud_ec2:describe_instances function
Implemented unit tests for it

Reviewers:
@ddosia @motobob

### CODE REVIEW

File: src/erlcloud_httpc.erl [hunk 1]
```diff
@@ -43,10 +43,8 @@ request(URL, Method, Hdrs, Body, Timeout,
     when is_function(F, 6) ->
     F(URL, Method, Hdrs, Body, Timeout, Config).
 
-request_lhttpc(URL, Method, Hdrs, Body, Timeout, #aws_config{lhttpc_pool = undefined}) ->
-    lhttpc:request(URL, Method, Hdrs, Body, Timeout, []);
-request_lhttpc(URL, Method, Hdrs, Body, Timeout, #aws_config{lhttpc_pool = Pool}) ->
-    lhttpc:request(URL, Method, Hdrs, Body, Timeout, [{pool, Pool}]).
+request_lhttpc(URL, Method, Hdrs, Body, Timeout, _Config) ->
```
Comment (motobob) [inline src/erlcloud_httpc.erl]: unnecessary change.

**Comment by motobob** [comment]
rebase on latest alertlogic/erlcloud will be appreciated and helpful for further merge

File: src/erlcloud_ec2.erl [hunk 2]
```diff
@@ -1347,36 +1349,133 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
+    when is_list(Filter), is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000, is_list(NextToken) ->
+    describe_instances(Filter, MaxResults, NextToken, default_config());
+describe_instances(Filter, MaxResults, Config)
+    when is_list(Filter), is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000, is_record(Config, aws_config) ->
+    describe_instances(Filter, MaxResults, undefined, Config).
+
+-spec describe_instances(filter_list(), ec2_max_result(), ec2_token(), aws_config()) -> ec2_paged_describe_return().
+describe_instances(Filter, MaxResults, NextToken, Config)
+    when is_list(Filter), is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000 orelse MaxResults =:= undefined, is_list(NextToken) orelse NextToken =:= undefined, is_record(Config, aws_config) ->
+    Params = list_to_ec2_filter(Filter) ++ [{"NextToken", NextToken}, {"MaxResults", MaxResults}],
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: when you will be implementing next funs you'll be surprised how much copy paster there will be so i still recommend considering updating erlcloud_util generic implementation or at leat separate it out in ec2 only ;)

File: src/erlcloud_ec2.erl [hunk 3]
```diff
@@ -1347,36 +1349,133 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
+                             (filter_list(), aws_config()) -> ec2_normal_describe_return();
+                             (ec2_max_result(), aws_config()) -> ec2_normal_describe_return().
+describe_instances_all(Filter, MaxResults)
+    when is_list(Filter), is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000 ->
+    describe_instances_all(Filter, MaxResults, undefined, default_config(), []);
+describe_instances_all(Filter, Config)
+    when is_list(Filter), is_record(Config, aws_config)->
+    describe_instances_all(Filter, 1000, undefined, Config, []);
+describe_instances_all(MaxResults, Config)
+    when is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000, is_record(Config, aws_config)->
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: stylish...use of both `andalso` and `,` stick to one way

File: src/erlcloud_ec2.erl [hunk 4]
```diff
@@ -1347,36 +1349,133 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
+describe_instances_all(Filter, MaxResults, Config)
+    when is_list(Filter), is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000, is_record(Config, aws_config)->
+    describe_instances_all(Filter, MaxResults, undefined, Config, []).
+
+-spec describe_instances_all(filter_list(), ec2_max_result(), ec2_token(), aws_config(), list()) -> ec2_normal_describe_return().
+describe_instances_all(Filter, MaxResults, NextToken, Config, Acc)
+    when is_list(Filter), is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000 orelse MaxResults =:= undefined, is_list(NextToken) orelse NextToken =:= undefined, is_record(Config, aws_config), is_list(Acc) ->
+    case describe_instances(Filter, MaxResults, NextToken, Config) of
+        {ok, Res, undefined} ->
+            {ok, Acc ++ Res};
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: is there any better way of concatenating a result?

File: test/erlcloud_ec2_tests.erl [hunk 5]
```diff
@@ -720,3 +720,172 @@ describe_images_tests(_) ->
...
+
+generate_describe_instances_response_header() ->
+    "<?xml version=\"1.0\" encoding=\"UTF-8\"?>
+    <DescribeInstancesResponse
+    xmlns=\"http://ec2.amazonaws.com/doc/2014-10-01/\">
+    <requestId>abcdef-12345</requestId>
+    <reservationSet>".
+
+generate_describe_instances_response_center() ->
+    "</reservationSet>".
```
Comment (motobob) [inline test/erlcloud_ec2_tests.erl]: this just make a dejavu with old school C macros
```C
#define BEGIN {
#define END }
```
which is very bad code style and to be considered for use only in auto generated code.
please remove

File: test/erlcloud_ec2_tests.erl [hunk 6]
```diff
@@ -720,3 +720,172 @@ describe_images_tests(_) ->
...
+
+generate_next_token_xml(NT) ->
+    NTString = integer_to_list(NT),
+    "<nextToken>" ++ NTString ++ "</nextToken>".
+
+generate_describe_instances_response_ending() ->
+    "</DescribeInstancesResponse>".
```
Comment (motobob) [inline test/erlcloud_ec2_tests.erl]: same as above

File: test/erlcloud_ec2_tests.erl [hunk 7]
```diff
@@ -720,3 +720,172 @@ describe_images_tests(_) ->
...
+                        true ->
+                            generate_instances_response(1, MaxInstances, undefined);
+                        false->
+                            generate_instances_response(1, MaxResults, MaxResults + 1)
+                    end;
+                NextToken ->
+                    NextTokenNumber = list_to_integer(NextToken),
+                    case NextTokenNumber + MaxResults - 1 >= MaxInstances of
+                        true ->
+                            generate_instances_response(NextTokenNumber, min(MaxInstances, NextTokenNumber + MaxResults - 1), undefined);
```
Comment (motobob) [inline test/erlcloud_ec2_tests.erl]: too long line.

File: src/erlcloud_ec2.erl [hunk 8]
```diff
@@ -1347,36 +1349,133 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
+            {ok, [extract_reservation(Item) || Item <- Reservations], NewNextToken};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+% Function for describing all instances page-by-page
+% Default page size is 1000
+-spec describe_instances_all() -> ec2_normal_describe_return().
+describe_instances_all() ->
+    describe_instances_all([], 1000, undefined, default_config(), []).
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: define 1000? 
or use one generic fun ;)

File: test/erlcloud_ec2_tests.erl [hunk 9]
```diff
@@ -720,3 +720,172 @@ describe_images_tests(_) ->
 
   %% Remaining AWS API examples return subsets of the same data
   output_tests(?_f(erlcloud_ec2:describe_images()), Tests).
+
+
+describe_instances_test() ->
+
```
Comment (motobob) [inline test/erlcloud_ec2_tests.erl]: stylish: whitespaces atack

File: test/erlcloud_ec2_tests.erl [hunk 10]
```diff
@@ -720,3 +720,172 @@ describe_images_tests(_) ->
...
+        fun(_,_,_,Params,_,_) ->
+            NextToken = proplists:get_value("NextToken",Params),
+            MaxResults = proplists:get_value("MaxResults",Params),
+            Response =
+            case NextToken of
+                undefined ->
+                    case MaxResults > MaxInstances of
+                        true ->
+                            generate_instances_response(1, MaxInstances, undefined);
+                        false->
```
Comment (motobob) [inline test/erlcloud_ec2_tests.erl]: and if `MaxResults =:= MaxInstances` ? 
please add a test
```erlang
describe_instances_pagination_tester(10, 10),
describe_instances_pagination_tester(0, 10),
```

File: test/erlcloud_ec2_tests.erl [hunk 11]
```diff
@@ -720,3 +720,172 @@ describe_images_tests(_) ->
...
+                                <code>code</code>
+                                <message>message</message>
+                            </stateReason>
+                            <architecture>x86_64</architecture>
+                            <rootDeviceType>ebs</rootDeviceType>
+                            <rootDeviceName>/dev/null</rootDeviceName>
+                            <blockDeviceMapping/>
+                            <virtualizationType>hvm</virtualizationType>
+                            <clientToken>123-123-12345-" ++ StringN ++ "</clientToken>
+                            " ++ generate_tags(5) ++ "
```
Comment (motobob) [inline test/erlcloud_ec2_tests.erl]: we already vary the data enough by `instanceID`. why do we need this?

File: test/erlcloud_ec2_tests.erl [hunk 12]
```diff
@@ -720,3 +720,172 @@ describe_images_tests(_) ->
...
+    xmlns=\"http://ec2.amazonaws.com/doc/2014-10-01/\">
+    <requestId>abcdef-12345</requestId>
+    <reservationSet>".
+
+generate_describe_instances_response_center() ->
+    "</reservationSet>".
+
+generate_next_token_xml(NT) ->
+    NTString = integer_to_list(NT),
+    "<nextToken>" ++ NTString ++ "</nextToken>".
```
Comment (motobob) [inline test/erlcloud_ec2_tests.erl]: string:concat() will do it for you nicely.

Comment (SergeyOvchinnik) [inline test/erlcloud_ec2_tests.erl]: Well, I see <code>concat()</code> actually uses <code>++</code> anyway, I just thought using ++ when adding several strings would be a bit more readable. Should I change it anyway?

File: src/erlcloud_ec2.erl (hunk 2 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: Agree, I'll implement that after I finish like 3 or 4 functions and know exactly which parts can vary between them

File: src/erlcloud_ec2.erl (hunk 4 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: Same as in comment below, as far as I can tell, other concatenating functions are equivalent to ++ but longer

Comment (ddosia) [inline src/erlcloud_ec2.erl]: Agree with @motobob. Since `Acc` is always growing in length and `Res` is +/- the same size, you will be moving all elements from big list to a smaller one. And correct me if I wrong, but it looks like you will change the order, i.e.:
```erlang
Acc = [1, 2, ... 100],
Res = [101, 102, ..., 200],
Acc ++ Res = [200,... 102, 101, 1, 2, ... 100].
```
So it looks like inefficient and wrong.
What I would do is compound a list of partially ordered lists, reverse it and flatten (i.e. `flatten(reverse([ResX..., Res2, Res1]))`.

Comment (motobob) [inline src/erlcloud_ec2.erl]: slight comment  - cannot flatten as this is a list of proplists.

Comment (motobob) [inline src/erlcloud_ec2.erl]: but in general let's rework this

Comment (ddosia) [inline src/erlcloud_ec2.erl]: Correction:
```erlang
Acc = [1, 2, ... 100],
Res = [101, 102, ..., 200],
Acc ++ Res = [100..., 2, 1, 101, 102, ..., 200].
```

File: src/erlcloud_ec2.erl [hunk 13]
```diff
@@ -1347,36 +1349,133 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
+    when is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000, is_record(Config, aws_config)->
+    describe_instances_all([], MaxResults, undefined, Config, []).
+
+-spec describe_instances_all(filter_list(), ec2_max_result(), aws_config()) -> ec2_normal_describe_return().
+describe_instances_all(Filter, MaxResults, Config)
+    when is_list(Filter), is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000, is_record(Config, aws_config)->
+    describe_instances_all(Filter, MaxResults, undefined, Config, []).
+
+-spec describe_instances_all(filter_list(), ec2_max_result(), ec2_token(), aws_config(), list()) -> ec2_normal_describe_return().
+describe_instances_all(Filter, MaxResults, NextToken, Config, Acc)
```
Comment (ddosia) [inline src/erlcloud_ec2.erl]: @motobob do you thing `describe_instances_all` is necessary? Since if I am not mistaken you can get all the data by invoking `describe_instances(Config)` function?

File: src/erlcloud_ec2.erl (hunk 4 above)
Comment (ddosia) [inline src/erlcloud_ec2.erl]: wait a second, this is bullshit. `++` does reverse inside, so everything is correct, but still inefficient :D

File: test/erlcloud_ec2_tests.erl [hunk 14]
```diff
@@ -720,3 +720,172 @@ describe_images_tests(_) ->
 
   %% Remaining AWS API examples return subsets of the same data
   output_tests(?_f(erlcloud_ec2:describe_images()), Tests).
+
+
+describe_instances_test() ->
+
+    describe_instances_pagination_tester(100, 20),
```
Comment (ddosia) [inline test/erlcloud_ec2_tests.erl]: It would be better if you convert this function to test generator (rename it to `describe_instances_test_` with underscore in the end), and generate three different test cases. Then I see in each `describe_instances_pagination_tester` you mock and unmock `erlcloud_aws` module. More right way to do this is by using eunit *setup* and *cleanup* fixtures.

Comment (ddosia) [inline test/erlcloud_ec2_tests.erl]: Why did you choose these numbers: `100, 20`, `50, 6`, `3, 5`? What are edge cases here?
Hint: test with page size within and out of bounds `5 =< X =< 1000`. When there zero pages, one, page, more than one page...
And write test description, what does it test and why.

File: test/erlcloud_ec2_tests.erl (hunk 12 above)
Comment (ddosia) [inline test/erlcloud_ec2_tests.erl]: For tests performance less matter than readability, imo.

**Comment by SergeyOvchinnik** [comment]
@motobob @ddosia I made some of the changes we discussed
